core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -934,10 +942,17 @@ private MergeWriterNode createMergePipeline(Table table, RelationPlan relationPl
                     columnNamesBuilder.add(columnSchema.getName());
                 });
         MergeParadigmAndTypes mergeParadigmAndTypes = new MergeParadigmAndTypes(Optional.of(paradigm), typesBuilder.build(), columnNamesBuilder.build(), rowIdType);
-        MergeTarget mergeTarget = new MergeTarget(handle, Optional.empty(), metadata.getTableName(session, handle).getSchemaTableName(), mergeParadigmAndTypes);
+        MergeTarget mergeTarget = new MergeTarget(handle, Optional.empty(), metadata.getTableName(session, handle).getSchemaTableName(), mergeParadigmAndTypes, updatedColumnHandles);
 
         ImmutableList.Builder<Symbol> columnSymbolsBuilder = ImmutableList.builder();
-        for (ColumnHandle columnHandle : mergeAnalysis.getDataColumnHandles()) {
+        List<ColumnHandle> iterateHandles;
```

Comment:
`updatedColumnHandles.isEmpty()` can be true also when handling a DELETE, no?
```suggestion
        List<ColumnHandle> columnsHandles = (paradigm != RowChangeParadigm.UPDATE_PARTIAL_COLUMNS) ? mergeAnalysis.getDataColumnHandles() : updatedColumnHandles;
```